internal/pkg/docker/dockercompose/volume.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package dockercompose
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	compose "github.com/compose-spec/compose-go/types"
+	"github.com/docker/go-units"
+	"github.com/dustin/go-humanize/english"
+	"strings"
+)
+
+// The default size that Copilot reserves for a service's ephemeral on-disk storage.
+const defaultEphemeralGiB = 20
+const defaultEphemeralBytes uint64 = defaultEphemeralGiB * units.GiB
+
+type volumeConverter struct {
+	otherSvcVols  map[string][]string
+	copilotVols   map[string]*manifest.Volume
+	topLevelVols  compose.Volumes
+	tmpfsVolNames map[string]bool
+}
+
+func newVolumeConverter(topLevelVols compose.Volumes) volumeConverter {
+	return volumeConverter{
+		otherSvcVols:  map[string][]string{},
+		copilotVols:   map[string]*manifest.Volume{},
+		topLevelVols:  topLevelVols,
+		tmpfsVolNames: map[string]bool{},
+	}
+}
+
+// convertVolumes converts a list of Compose volumes into the Copilot storage equivalent.
+func (vc *volumeConverter) convertVolumes(volumes []compose.ServiceVolumeConfig, otherSvcs compose.Services) (*manifest.Storage, IgnoredKeys, error) {
+	for _, otherSvc := range otherSvcs {
+		for _, vol := range otherSvc.Volumes {
+			if vol.Type == "volume" {
+				vc.otherSvcVols[vol.Source] = append(vc.otherSvcVols[vol.Source], otherSvc.Name)
+			}
+		}
+	}
+
+	var ephemeralBytes = defaultEphemeralBytes
+	var ignored IgnoredKeys
+
+	for idx, vol := range volumes {
+		if vol.Target == "" {
+			return nil, nil, fmt.Errorf("volume mounted from %q (type %q) is missing a target mount point", vol.Source, vol.Type)
+		}
+
+		mountOpts := manifest.MountPointOpts{
+			ContainerPath: aws.String(vol.Target),
+			ReadOnly:      aws.Bool(vol.ReadOnly),
+		}
+
+		if vol.Type == "tmpfs" {
+			ephemeralBytes += uint64(vol.Tmpfs.Size)
```

Comment:
i don't know much about volumes in compose so i might just not understand something - but it seems like the logic is to:
a. default to 20gb 
b. if tmpfs, add 20gb to whatever the sum of configured volumes is

and i'm curious about option b there - why are we adding 20 gb no matter what they specify?

Reply:
That analysis is correct! It's an approximation of Compose behavior because the only relevant meaning of a tmpfs mount that matters to Copilot is if the service needs it for additional temporary storage.

I chose this approach, since if for example the service mounted a 100mb tmpfs volume as extra space then we wouldn't want to reduce the total ephemeral storage to 100MB. But, if it mounts a 50 GB tmpfs volume, then certainly we'd need that amount of additional space. I think this approach is the most sensible sensible since in either case, this method will ensure that the needs of the service are met.